### PR TITLE
Produce semi shortest path

### PR DESCRIPTION
**Pull request by nevermore3**

### CODE REVIEW

File: src/executor/algo/ProduceSemiShortestPathExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            auto oldCost = currentCostPathMap[dst][srcPath.first].first;
+            if (newCost > oldCost) {
+                continue;
+            } else if (newCost < oldCost) {
+                // update (dst->src)'s path
+                std::vector<Path> newPaths = createPaths(srcPath.second.second, edge);
+                currentCostPathMap[dst][srcPath.first].first = newCost;
+                currentCostPathMap[dst][srcPath.first].second.swap(newPaths);
+            } else {
+                // add (dst->src)'s path
```
Comment (CPWstatic) [inline src/executor/algo/ProduceSemiShortestPathExecutor.cpp:46]: What do you mean `dst->src`? Maybe
```
path + (src, dst) 
```

File: src/executor/algo/ProduceSemiShortestPathExecutor.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    ProduceSemiShortestPathExecutor(const PlanNode* node, QueryContext* qctx)
+        : Executor("ProduceSemiShortestPath", node, qctx) {}
+
+    folly::Future<Status> execute() override;
+
+    using CostPathMapType =
+        std::unordered_map<Value, std::unordered_map<Value, std::pair<int64_t, std::vector<Path>>>>;
+    using CostPathMapPtr =
+        std::unordered_map<Value,
+                           std::unordered_map<Value, std::pair<int64_t, std::vector<Path*>>>>;
```
Comment (CPWstatic) [inline src/executor/algo/ProduceSemiShortestPathExecutor.h:25]: Need some comments.
Suggest using struct instead pair.

File: src/executor/algo/ProduceSemiShortestPathExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+                                                   CostPathMapType& currentCostPathMap) {
+    auto& src = edge.src;
+    auto& dst = edge.dst;
+    auto weight = 1;   // weight = weight_->getWeight();
+    auto& srcPaths = historyCostPathMap_[src];
+
+    for (auto& srcPath : srcPaths) {
+        if (currentCostPathMap[dst].find(srcPath.first) != currentCostPathMap[dst].end()) {
+            // src in currentCostPathMap[dst]
+            auto newCost = srcPath.second.first + weight;
```
Comment (CPWstatic) [inline src/executor/algo/ProduceSemiShortestPathExecutor.cpp:36]: For further concern, i suggest that add a method to calculate the cost.

File: src/executor/algo/ProduceSemiShortestPathExecutor.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            if (historyCostPathMap_.find(dst) == historyCostPathMap_.end()) {
+                dstNotInHistory(edge, currentCostPathMap);
+            } else {
+                dstInHistory(edge, currentCostPathMap);
+            }
+        }
+    }
+
+    DataSet ds;
+    ds.colNames = node()->colNames();
```
Comment (CPWstatic) [inline src/executor/algo/ProduceSemiShortestPathExecutor.cpp:167]: Please fulfill the dataset and update the history map.

File: src/executor/algo/ProduceSemiShortestPathExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+namespace nebula {
+namespace graph {
+
+std::vector<Path> ProduceSemiShortestPathExecutor::createPaths(const std::vector<Path*>& paths,
+                                                               const Edge& edge) {
+    std::vector<Path> newPaths;
+    newPaths.reserve(paths.size());
+    for (auto p : paths) {
+        Path path = *p;
+        path.steps.emplace_back(Step(Vertex(edge.dst, {}), edge.type, edge.name, edge.ranking, {}));
```
Comment (CPWstatic) [inline src/executor/algo/ProduceSemiShortestPathExecutor.cpp:20]: Maybe we could using `iter->getVertex()` to get the vertices directly.

**Review by CPWstatic — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**